<?xml version="1.0"?>
<launch>

    <arg name="model" default="$(find sq2_ccv_description)/robot/sq2_ccv.urdf"/>

    <arg name="paused" default="false"/>
    <arg name="use_sim_time" default="true"/>
    <arg name="gui" default="true"/>
    <arg name="headless" default="false"/>
    <arg name="debug" default="false"/>
    <arg name="x" default="0.0"/>
    <arg name="y" default="0.0"/>
    <arg name="z" default="0.0"/>
    <arg name="roll" default="0.0"/>
    <arg name="pitch" default="0.0"/>
    <arg name="yaw" default="0.0"/>

    <include file="$(find gazebo_ros)/launch/empty_world.launch">
        <arg name="debug" value="$(arg debug)" />
        <arg name="gui" value="$(arg gui)" />
        <arg name="paused" value="$(arg paused)"/>
        <arg name="use_sim_time" value="$(arg use_sim_time)"/>
        <arg name="headless" value="$(arg headless)"/>
    </include>

    <param name="robot_description" command="$(find xacro)/xacro $(arg model)"/>

    <include file="$(find sq2_ccv_description)/launch/control.launch"/>

    <node name="urdf_spawner" pkg="gazebo_ros" type="spawn_model" respawn="false" output="screen"
        args="-urdf
        -model sq2_ccv
        -param robot_description
        -x $(arg x)
        -y $(arg y)
        -z $(arg z)
        -R $(arg roll)
        -P $(arg pitch)
        -Y $(arg yaw)"/>

    <node pkg="rostopic" type="rostopic" name="init_roll" args="pub -1 /sq2_ccv/battery_box_joint/command std_msgs/Float64 'data: 0.0'"/>
    <node pkg="rostopic" type="rostopic" name="init_left_steering" args="pub -1 /sq2_ccv/left_steering/command std_msgs/Float64 'data: 0.0'"/>
    <node pkg="rostopic" type="rostopic" name="init_right_steering" args="pub -1 /sq2_ccv/right_steering/command std_msgs/Float64 'data: 0.0'"/>

    <node name="test_path_generator" pkg="steering_path_planner" type="test_path_creator">
        <!-- <param name="world_frame_id" value="world"/> -->
        <param name="world_frame_id" value="odom"/>
        <param name="resolution" value="0.05"/>
        <param name="course_length" value="7.0"/>
        <param name="init_x" value="2.0"/><!-- -4.957 -->
        <param name="init_y" value="0.3"/><!-- 1.886 -->
        <param name="init_theta" value="0.75"/><!-- -0.355 -->
        <param name="A1" value="0.5"/>
        <param name="omega1" value="85"/>
        <!-- <param name="omega1" value="85"/> -->
        <param name="delta1" value="220"/>
        <param name="A2" value="0.0"/>
        <param name="omega2" value="0"/>
        <param name="delta2" value="0"/>
        <param name="A3" value="0.0"/>
        <param name="omega3" value="0"/>
        <param name="delta3" value="0"/>
    </node>


    <node name="diff_drive_mpc" pkg="ccv_mpc_steering" type="diff_drive_mpc" output="screen">
        <remap from="/local/cmd_vel" to="/sq2_ccv/diff_drive_steering_controller/cmd_vel"/>
    </node>

    <node name="steering_angle_publisher" pkg="ccv_pure_pursuit_steering" type="steering_angle_publisher"/>

    <node name="rviz" pkg="rviz" type="rviz" args="-d $(find ccv_mpc_steering)/config/diff_drive_mpc.rviz"/>
</launch>
